### ISSUE TITLE
Incorrect cell suppression with specific example

### ISSUE THREAD

**Issue by PeterM74** [inpowell/modulartabler#17]
Hope all is going well with you with your recent movements! I will attempt to tackle this one - it appears that the cell suppression algorithm has some instances where it doesn't appear to be suppressing all possible backwards calculations. I have included an example below, with variables recoded and Ns shifted bidirectionally while still maintaining the core relationships that are triggering the bug. 

``` r
devtools::load_all(".")
#> ℹ Loading modulartabler
  
  compactform <- structure(list(
    `Factor1` = structure(c(1L, 1L, 1L, 2L, 2L, 2L, 3L, 3L, 
                            3L, 4L, 4L, 4L, 5L, 5L, 5L, 6L, 6L, 6L,
                            7L, 7L, 7L, 8L, 8L, 8L, 9L, 9L, 9L),
                          levels = c("Lvl1", "Lvl2", "Lvl3", "Lvl4", 
                                     "Lvl5", "Lvl6", "Lvl7", "Lvl8", 
                                     "Total"), class = "factor"),
    `Factor2` = structure(c(1L, 2L, 3L, 1L, 2L, 3L, 1L, 2L, 3L, 1L, 
                            2L, 3L, 1L, 2L, 3L, 1L, 2L,
                            3L, 1L, 2L, 3L, 1L, 2L, 3L, 1L, 2L, 3L), 
                          levels = c("Level1", "Level2", "Total"), 
                          class = "factor"),
    n = c(262L, 78L, 340L, 110L, 59L, 169L, 52L, 16L, 68L, 54L, 16L, 
          70L, 97L, 17L, 114L, 90L, 37L, 127L, 1L, 3L, 4L, 0L, 0L, 
          0L, 666L, 226L, 892L)),
    row.names = c(NA, -27L), class = c("tbl_df", "tbl", "data.frame"))
  
  Fct1Map <- RangeMappingTable$new(
    "Factor1", "fct1",
    "Lvl1" = c(-Inf, 1),
    "Lvl2" = c(1, 3),
    "Lvl3" = c(3, 5),
    "Lvl4" = c(5, 7),
    "Lvl5" = c(7, 9),
    "Lvl6" = c(9, 11),
    "Lvl7" = c(11, Inf),
    .other = "Lvl8",
    .total = "Total"
  )
  
  Fct2Map <- BaseMappingTable$new(
    tibble::tibble(fct2 = rep(1:2, 2),
                   Factor2 = forcats::as_factor(c("Level1", "Level2", rep("Total", 2)))),
    "fct2", "Factor2"
  )
  
  MMap <- MultiMappingTable$new(Fct1Map, Fct2Map)
  
  test_suppress <- compactform %>%
    dplyr::mutate(suppress = determine_cell_suppression(n, MMap$nullspace))
  
  convert_tabular(test_suppress, Factor1 ~ Factor2 + .Measure, 
                  measures = c(No. = "n", S = "suppress"))
#> # A tibble: 9 × 7
#>   Factor1 Level1_No. Level1_S Level2_No. Level2_S Total_No. Total_S
#>   <fct>        <int> <lgl>         <int> <lgl>        <int> <lgl>  
#> 1 Lvl1           262 FALSE            78 FALSE          340 FALSE  
#> 2 Lvl2           110 FALSE            59 FALSE          169 FALSE  
#> 3 Lvl3            52 TRUE             16 FALSE           68 TRUE   
#> 4 Lvl4            54 FALSE            16 FALSE           70 FALSE  
#> 5 Lvl5            97 FALSE            17 FALSE          114 FALSE  
#> 6 Lvl6            90 FALSE            37 FALSE          127 FALSE  
#> 7 Lvl7             1 TRUE              3 TRUE             4 TRUE   
#> 8 Lvl8             0 FALSE             0 FALSE            0 FALSE  
#> 9 Total          666 FALSE           226 FALSE          892 FALSE
  
  # For messing around with numbers
  raw_data <- compactform %>%
    dplyr::filter(dplyr::if_all(c(Factor1, Factor2), ~ .x != "Total")) %>%
    tidyr::uncount(weights = n)
```

<sup>Created on 2024-09-07 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

As you can see for the Level2 column, only one value has been suppressed so the Lvl7 row can be recalculated off this total. I have some theories for what is triggering this but will delve into this further.

**Comment by inpowell** [comment]
Ooh that’s interesting! I am without computer for the next few weeks probably, so if you could take a look that would be great.

If I did have one, I would first check that the nullspace is correct. (I think I would expect it to have 11 rows: 9 row totals + 3 column totals - 1 duplicated grand total.) If not, then I’d have a very exciting delve into suppress_secondary()!

Also very interested to hear your thoughts on what’s causing this.

**Comment by PeterM74** [comment]
I figured! Yes I will look into this further. I too thought there was an issue with the nullspace (although different to what you were thinking, I thought the grand total was interferring with the columnar nullspace representation), but unfortunately manually mapping the nullspace didn't fix it so it wasn't that! You are correct that there are 11 rows and reviewing the specification of each row of the nullspace didn't reveal any issue. Row 10 of the matrix perfectly represents the column with the issue so unfortunately it does look like I'm going to have to go deep into suppress_secondary which I was really hoping to avoid! Will let you know what I find.

**Comment by inpowell** [comment]
Does this problem persist if you use a different solver? HiGHS is the default because it seemed quickest when I ran some tests, but it might be interesting to see if lpSolve (or some other solver) gives the correct answer

**Comment by PeterM74** [comment]
Unfortunately I ran the example on the old version that used lpSolve and it outputs the same solution. ~~I wonder if there is a bug in the definition of the problem (that has somehow not triggered before now). I wouldn't think the solvers would have a problem (especially HiGHS which is quite popular) but are finding the 'solution' to a misspecified dimension of the problem?~~ After reading more on the solvers, I realise the solver can't be ruled out although lpSolver and HiGHS both arrive at the same answer in this instance.

**Update**:
I output the candidate suppression after running it over the 3 primary suppressed cells (rows 19/20/21). Unsurprisingly, when the master problem only considers row/cell 20 (the cell that is vulnerable to the attack in this example), the algorithm correctly secondary suppresses row 8 (and row 7 correctly). When cells 19 & 20 constraints are added in sequence (but not 21), the correct solution to that point is still output (solution to 19 also solves 20). It is only when cell 21 is included in the mix that suddenly row 8 loses it's suppression when the additional constraint is added from 21. Somehow I guess the constraints added from cell 21 is outweighing previously added constraints? Investigations continue...
```
# A tibble: 27 x 8
   Factor1 Factor2     n suppress suppress_only_row19 suppress_only_row20 suppress_only_row21 suppress_both_1920
   <fct>   <fct>   <int> <lgl>    <lgl>               <lgl>               <lgl>               <lgl>             
 1 Lvl1    Level1    262 FALSE    FALSE               FALSE               FALSE               FALSE             
 2 Lvl1    Level2     78 FALSE    FALSE               FALSE               FALSE               FALSE             
 3 Lvl1    Total     340 FALSE    FALSE               FALSE               FALSE               FALSE             
 4 Lvl2    Level1    110 FALSE    FALSE               FALSE               FALSE               FALSE             
 5 Lvl2    Level2     59 FALSE    FALSE               FALSE               FALSE               FALSE             
 6 Lvl2    Total     169 FALSE    FALSE               FALSE               FALSE               FALSE             
 7 Lvl3    Level1     52 TRUE     TRUE                TRUE                FALSE               TRUE              
 8 Lvl3    Level2     16 FALSE    TRUE                TRUE                TRUE                TRUE              
 9 Lvl3    Total      68 TRUE     FALSE               FALSE               TRUE                FALSE             
10 Lvl4    Level1     54 FALSE    FALSE               FALSE               FALSE               FALSE             
11 Lvl4    Level2     16 FALSE    FALSE               FALSE               FALSE               FALSE             
12 Lvl4    Total      70 FALSE    FALSE               FALSE               FALSE               FALSE             
13 Lvl5    Level1     97 FALSE    FALSE               FALSE               FALSE               FALSE             
14 Lvl5    Level2     17 FALSE    FALSE               FALSE               FALSE               FALSE             
15 Lvl5    Total     114 FALSE    FALSE               FALSE               FALSE               FALSE             
16 Lvl6    Level1     90 FALSE    FALSE               FALSE               FALSE               FALSE             
17 Lvl6    Level2     37 FALSE    FALSE               FALSE               FALSE               FALSE             
18 Lvl6    Total     127 FALSE    FALSE               FALSE               FALSE               FALSE             
19 Lvl7    Level1      1 TRUE     TRUE                TRUE                TRUE                TRUE              
20 Lvl7    Level2      3 TRUE     TRUE                TRUE                TRUE                TRUE              
21 Lvl7    Total       4 TRUE     TRUE                TRUE                TRUE                TRUE              
22 Lvl8    Level1      0 FALSE    FALSE               FALSE               FALSE               FALSE             
23 Lvl8    Level2      0 FALSE    FALSE               FALSE               FALSE               FALSE             
24 Lvl8    Total       0 FALSE    FALSE               FALSE               FALSE               FALSE             
25 Total   Level1    666 FALSE    FALSE               FALSE               FALSE               FALSE             
26 Total   Level2    226 FALSE    FALSE               FALSE               FALSE               FALSE             
27 Total   Total     892 FALSE    FALSE               FALSE               FALSE               FALSE 
```

**Comment by PeterM74** [comment]
Oh my God, I think I have found it after about 5hrs of reading the article, reading the code, executing the code step by step and cursing my existence. 

<img src="https://media1.tenor.com/m/lnjI7nu2tGgAAAAd/ron-swanson-throws-out-his-computer.gif" height="200" />

I added a logging system within the loops so I could see what cells were adding which subproblems to the constraints which I think has solved it. Cell 19 added 3 constraints and candidate suppression solved itself correctly. Cell 20 was conveniently covered by the candidate_suppression from cell 19 so **didn't** add a constraint (this will come back to bite us). Cell 21 adds a constraint and the loop ends because cell 21 passed but because cell 20 didn't add a constraint, only cells 19 and 21 had constraints added to the model so they are the only considerations. Clearly while cell 19s constraints *by themselves* were sufficient to cover cell 20, the addition of cell 21 changed the balance. The solver is working perfectly, the issue is that one cell wasn't considered! Running the loop a second time with the cell 21 constraints ended up adding a constraint to cell 20 on the second passthrough and rendered the table (properly) solved!

Possible solutions:

* Whenever the `candidate_suppression` is updated, this is checked against all primary suppressed cells again (i.e. in the `attack.ik` loop). At the moment the `candidate_suppression` is updated as we loop through the primary suppressed cells in `attack.ik` and there is a risk that certain suppression patterns will cover future cells but these aren't explicitly added to the constraints so are at risk for future changes. Because the constraints haven't been added and were assumed solved by previous constraints, future constraint updates may outweigh the serendipitous coverage these cells had from other cell's constraints (if that makes sense).
* Rather than use the 'global' `candidate_suppression` in the calculation of `attacker.max` and `attacker.min`, could we use a local `candidate_suppression` specific to that cell? So each cell will calculate it's own candidate_suppression until it finally passes all attacker subproblems, then the `master_lp` is solved at the end, combining all constraints?

I would really appreciate your insight into this and if you have any alternate suggestions or thoughts on preferred solution! Hope this makes sense, happy to take a Zoom/Teams call after you've settled in if my explanation above doesn't make sense!

**Comment by inpowell** [comment]
Woah... nice work and that makes sense to me as the issue.

I’ve just gone back to read the paper again (specifically sec 4.1 “After having considered _all_ sensitive cells...”, emphasis added). From that reading and what you’ve written, I think that the logical error in the current implementation is that we are solving the master LP after each cell attack. However, what we _want_ to do would be solve the master LP after adding constraints for all sensitive cells.

Looking at the implementation, I think (in rough terms, with no testing capability on my end) we would want to flip the for-loops so they go:

```r
  i <- 1L
  repeat { # Master LP loop
    attacker_success <- FALSE # moved from inner loop to here
    for (attack.ik in ik) { # Attack all cells to be suppressed
      ...
    }

    # If there is no successful avenue of attack, we are done!
    if (!attacker_success) break

    # If we have tried too hard, give up
    if (identical(i, max_iter) && attacker_success) {
      stop("Maximum attacker iterations reached")

    # Re-solve master LP with new constraints
    i <- i + 1L
    master_soln <- ROI_solve(master_lp, ...)
    candidate_suppression <- as.logical(master_soln$solution)
}
```

In this thinking, max_iter corresponds to the number of times we’re happy to solve the master LP (and how many times we’re willing to attack each sensitive cell in total). I think that makes more sense to me as a parameter.

(In theory, I think one could parallelise the attacker subproblems. In practice, my guess would be that it wouldn’t be worth it in R, and would need low-level C/C++.)

**Comment by PeterM74** [comment]
Great! I have implemented your suggestion and it is passing the tests and the above example 🙌 May try a few more and also do some benchmarking to test it out. Here's a question - when we do `for (attack.ik in ik) ...`, do we need to include the newly suppressed cells in candidate_suppression in `ik` to to make sure they are also properly suppressed and resistant to back calculation? 

~~Edit: I just looked at the answer to the above example and one of the new secondary suppressions can be back calculated so I believe they do. Will do some testing. Hmmm, but it's adding non-optimal solutions to other tables... perhaps there's an issue with my code...~~ 

Never mind, there was an error in my code! All tests passing, example solved for above. It is more computationally more expensive now as it's needing to loop through things more often. Will think of any optimisations we can include on our end, but agree that we can leave parallelism to a future update!

**Comment by inpowell** [comment]
My quite-firm belief would be that only primary suppressed cells need to be attacked. From my reading of the paper, it treats the set of $i_k$ as immutable. Also, if we treated secondary suppressed cells as requiring suppression, I think this would lead to redundant suppressions.

From a theoretical point of view, we want the suppression pattern that minimises suppressed weight (i.e. `w`) subject to keeping sensitive cells safe. If we start requiring suppression of non-sensitive cells, this would (likely) lead to a suboptimal pattern with a higher suppressed weight.

The final solution of the master LP gets attacked for all sensitive cells (in the new implementation), so we know it’s safe (or else errors eventually).

**Comment by PeterM74** [comment]
You are right - I think I needed to sleep on it as the continual attack loop should sort out downstream issues 😄 The current implementation should be sufficient. Next steps for me is:

- [ ] Conduct further testing: I might start using this implementation in our internal work to test it out as there are a lot of tables being generated I can test it on
- [ ] Conduct some benchmarking on some highly dimensional tables so the hit. It was quite performant in the examples I ran but they were all very simple
- [x] I might add the above example as a unit test for suppression. I don't think it is likely that we will regress on this front, but I think it is an interesting example where the constraints for individual cells can be tricky, where if we don't conduct a final loop of the suppressed cells, we run into trouble with the stacking of constraints.

I will then submit a PR but understand it may be a few weeks until you can review properly. No rush or pressure to review when I do!

---
Benchmarking update 1 (2024/09/11):
As expected, there has been a performance hit as rather than looping once through all primary suppressed cells, there is at least 2 loops (and possibly more). Running 6 identical tables (3 x 2 dimensions, 3 x 3 dimensions) across the existing and new algos:

```
   algo  version                        `median (s)` Cumulative_mem  n_gc
   <chr> <chr>                                 <dbl>      <bch:byt> <dbl>
 1 Old   Maps: 2, items: 10, version: 1        0.505        68.99MB   141
 2 Old   Maps: 2, items: 10, version: 2        0.492        64.28MB   115
 3 Old   Maps: 2, items: 10, version: 3        0.482        60.71MB   115
 4 Old   Maps: 3, items: 5, version: 1         7.11        973.67MB   994
 5 Old   Maps: 3, items: 5, version: 2         8.90        824.64MB   823
 6 Old   Maps: 3, items: 5, version: 3         9.73        973.57MB   986
-------------------------------------------------------------------------
 7 New   Maps: 2, items: 10, version: 1        1.22        175.04MB   232
 8 New   Maps: 2, items: 10, version: 2        0.714        99.51MB   130
 9 New   Maps: 2, items: 10, version: 3        0.885       123.41MB   163
10 New   Maps: 3, items: 5, version: 1         47.5           2.15GB  1658
11 New   Maps: 3, items: 5, version: 2         67.0           1.97GB  1534
12 New   Maps: 3, items: 5, version: 3         19.1           1.47GB  1141
-------------------------------------------------------------------------
13 New2  Maps: 2, items: 10, version: 1        0.747       117.16MB   184
14 New2  Maps: 2, items: 10, version: 2        0.489        72.15MB   104
15 New2  Maps: 2, items: 10, version: 3        0.618        90.14MB   131
16 New2  Maps: 3, items: 5, version: 1         4.29          1.19GB  1144
17 New2  Maps: 3, items: 5, version: 2         7.84          1.77GB  1703
18 New2  Maps: 3, items: 5, version: 3         5.51          1.44GB  1398
```
`Cumulative_mem` is the total cumulative read/write to memory and I believe this is the cause of the significant time differences (200-700% increase). Specifically, the number of calls to the garbage collector is concerning and unsurprising for the looping mechanism employed. This is likely causing a lot of reallocations, knowing how R does copy on modify and doesn't modify in place. I already have some ideas to optimise (there is a lot of redundant checking that can be easily eliminated). I also have a feeling the current implementation I have done is unneeded - I don't think I need to be using the most up to date candidate suppression and can get away with adding the constraints as you have done and then solving the master lp ... I am not sure why I did code it this way! On to solution 2...

On another note, I did encounter another bug that only triggers on 4 dimension tables and higher (so far, probably affects very large 3 dimensional tables as well) - an error is triggering on tables that really aren't very computationally expensive I would think (e.g. 4 dimensional table where each table has only two elements plus total). I might submit a second bug but this doesn't need to be solved before releasing on CRAN as it throws an error and is only triggered on what I think is a rare table formulation).

---
Benchmarking update 2 (2024/09/12):
Second iteration of algorithm added to above. Similar performance to original algorithm.

**Comment by PeterM74** [comment]
Regrettably I seem to have encountered another bug while I was doing extensive testing 🤦‍♂️ I was so excited thinking we'd finally nailed it! Better to get them out of the way now than later I guess! This bug is replicable across both the original algo, the second (overly thorough and unperformant) algo, and now the latest one. See steps below to reproduce and result (3 dimensional table with 4 elements each), note some of the subgroupings I've asterisked:

```
# Coded this way as I was outputting many tables to test and benchmark so wanted to produce quickly
f <- function(colname, max = 5) {
  BaseMappingTable$new(tibble::tibble(!!colname := c(letters[1:max],
                                                     letters[1:max]),
                                      !!paste0(colname, "2") := c(LETTERS[1:max],
                                                                  rep("Total", max))),
                       colname, paste0(colname, "2"))
}

map <- do.call(MultiMappingTable$new, purrr::map(LETTERS[1:3], ~ f(.x, max = 4)))

aggregate <- do.call(tidyr::expand_grid, map(1:3, ~ letters[1:4])) %>%
    purrr::set_names(nm = LETTERS[1:3])

set.seed(157)
raw <- dplyr::mutate(aggregate, n = sample(0L:40L, size = 4^3, replace = TRUE)) %>%
    tidyr::uncount(weights = n)

final_aggregate <- map$count_aggregate(raw) %>% 
    dplyr::mutate(output = determine_cell_suppression(n, map$nullspace))

convert_tabular(final_aggregate, A2 + B2 ~ C2 + .Measure, measures = c(No. = "n", s = "output")) %>% 
    print(n = 25)

   A2    B2    A_No. A_s   B_No. B_s   C_No. C_s   D_No. D_s   Total_No. Total_s
   <chr> <chr> <int> <lgl> <int> <lgl> <int> <lgl> <int> <lgl>     <int> <lgl>  
 1 A     A         1 TRUE     35 FALSE    37 FALSE    31 TRUE        104 FALSE  
 2 A     B         2 TRUE     36 FALSE    27 FALSE     9 FALSE        74 FALSE  
 3 A     C        38 FALSE    37 FALSE    24 FALSE    29 TRUE        128 FALSE  
 4 A     D        22 TRUE     25 FALSE    35 FALSE     2 TRUE         84 FALSE  
 5 A     Total    63 FALSE   133 FALSE   123 FALSE    71 TRUE        390 FALSE  
 6 B     A        40 TRUE     30 FALSE    31 FALSE     6 TRUE        107 FALSE  
 7 B     B         3 TRUE     39 FALSE    28 TRUE     24 FALSE        94 FALSE  
 8 B     C        32 TRUE     15 TRUE     18 TRUE      2 TRUE         67 FALSE  
 9 B     D        35 FALSE    10 TRUE     10 TRUE     32 TRUE         87 FALSE  
10 B     Total   110 FALSE    94 FALSE    87 FALSE    64 FALSE       355 FALSE  
11 C     A        21 FALSE*    6 FALSE     9 FALSE    24 TRUE         60 FALSE  
12 C     B        16 FALSE*   36 FALSE    30 FALSE    23 FALSE       105 FALSE  
13 C     C        39 FALSE*   22 FALSE    19 FALSE    35 FALSE       115 FALSE  
14 C     D         9 TRUE*    16 FALSE    17 FALSE    37 FALSE        79 FALSE  
15 C     Total    85 FALSE*   80 FALSE    75 FALSE   119 FALSE       359 FALSE  
16 D     A        33 FALSE*   19 FALSE    34 FALSE*   33 FALSE       119 FALSE  
17 D     B        10 FALSE*   24 FALSE    29 FALSE*   25 FALSE        88 FALSE  
18 D     C        36 FALSE*    6 TRUE     26 TRUE*    17 TRUE         85 FALSE  
19 D     D        22 TRUE*    29 TRUE     26 FALSE*   16 TRUE         93 FALSE  
20 D     Total   101 FALSE*   78 FALSE   115 FALSE*   91 FALSE       385 FALSE  
21 Total A        95 FALSE    90 FALSE   111 FALSE    94 FALSE       390 FALSE  
22 Total B        31 FALSE   135 FALSE   114 FALSE    81 FALSE       361 FALSE  
23 Total C       145 FALSE    80 FALSE    87 FALSE    83 FALSE       395 FALSE  
24 Total D        88 FALSE    80 FALSE    88 FALSE    87 FALSE       343 FALSE  
25 Total Total   359 FALSE   385 FALSE   400 FALSE   345 FALSE      1489 FALSE  
```
These are all secondary suppressions that haven't been suppressed further to prevent back-calculation. Given this issue is replicable across all 3 algorithms (original, first iteration, latest iteration), I'm thinking there must be something going on somewhere. Haven't found anything immediately obvious but will keep investigating...

~~**Update**
Ok, but a hold on this for now. I ran with 'glpk' and it came to the correct solution. I then ran with HiGHs and it solved it correctly this time... I'm too tired for this 🤣 Potentially I'm not running with a clean environment or something funky is going on. Will try again tomorrow~~

**Comment by PeterM74** [comment]
**Update: this has been partially resolved! See next comment but worth a read for context and for the potential cache issues in ROI**

Hi @inpowell, might need your insight on this one when you have time... I am encountering a very strange bug. I am getting consistently inconsistent results, as you will see! When I run the HiGHS solver, I get a suboptimal solution. When I run the GLPK solver, I get (what appears to be) the optimal solution. When I run the HiGHS solver *after* GLPK, I get the optimal solution. So I think the code is working and there is something very odd happening in the underlying ROI package or underlying solvers.

I have restarted my computer 3 times to completely clear any memory in case of cache/memory issues and to have a clear R (overkill I know but this is really annoying me), I don't have any unexpected global variables or anything that shouldn't be there within the environments I can see. Here's the code causing issues:

```
devtools::load_all()
library(tictoc)

# Compact way to build out all the mappings for this 3 dimensional mapping
f <- function(colname, max = 5) {
  BaseMappingTable$new(tibble::tibble(!!colname := c(letters[1:max],
                                                     letters[1:max]),
                                      !!paste0(colname, "2") := c(LETTERS[1:max],
                                                                  rep("Total", max))),
                       colname, paste0(colname, "2"))
}

map <- do.call(MultiMappingTable$new, purrr::map(LETTERS[1:3], ~ f(.x, max = 4)))

aggregate <- do.call(tidyr::expand_grid, map(1:3, ~ letters[1:4])) %>%
  purrr::set_names(nm = LETTERS[1:3])

set.seed(157)
raw <- dplyr::mutate(aggregate, n = sample(0L:40L, size = 4^3, replace = TRUE)) %>%
  tidyr::uncount(weights = n)

# Default HiGHS implementation
## This takes 3.64s - perfect!
tic() 
final_aggregate <- map$count_aggregate(raw) %>%
  dplyr::mutate(highs_1 = determine_cell_suppression(n, map$nullspace))
toc() 

convert_tabular(final_aggregate, A2 + B2 ~ C2 + .Measure, measures = c(No. = "n", s = "highs_1")) %>%
  print(n = 25)

# Very unoptimised solution - way too much suppression implemented and some can be backcalculated
   A2    B2    A_No. A_s   B_No. B_s   C_No. C_s   D_No. D_s   Total_No. Total_s
   <chr> <chr> <int> <lgl> <int> <lgl> <int> <lgl> <int> <lgl>     <int> <lgl>  
 1 A     A         1 TRUE     35 TRUE     37 FALSE    31 TRUE        104 FALSE  
 2 A     B         2 TRUE     36 FALSE    27 FALSE     9 TRUE         74 FALSE  
 3 A     C        38 FALSE    37 FALSE    24 FALSE    29 TRUE        128 FALSE  
 4 A     D        22 TRUE     25 TRUE     35 TRUE      2 TRUE         84 FALSE  
 5 A     Total    63 FALSE   133 FALSE   123 FALSE    71 TRUE        390 FALSE  
 6 B     A        40 TRUE     30 TRUE     31 FALSE     6 TRUE        107 FALSE  
 7 B     B         3 TRUE     39 FALSE    28 FALSE    24 TRUE         94 FALSE  
 8 B     C        32 TRUE     15 TRUE     18 TRUE      2 TRUE         67 FALSE  
 9 B     D        35 TRUE     10 FALSE    10 TRUE     32 TRUE         87 FALSE  
10 B     Total   110 FALSE    94 FALSE    87 FALSE    64 TRUE        355 FALSE  
11 C     A        21 FALSE     6 TRUE      9 FALSE    24 FALSE        60 FALSE  
12 C     B        16 TRUE     36 FALSE    30 FALSE    23 FALSE       105 FALSE  
13 C     C        39 FALSE    22 FALSE    19 FALSE    35 FALSE       115 FALSE  
14 C     D         9 TRUE     16 FALSE    17 FALSE    37 FALSE        79 FALSE  
15 C     Total    85 FALSE    80 FALSE    75 FALSE   119 FALSE       359 FALSE  
16 D     A        33 FALSE    19 TRUE     34 FALSE    33 FALSE       119 FALSE  
17 D     B        10 TRUE     24 FALSE    29 FALSE    25 FALSE        88 FALSE  
18 D     C        36 FALSE     6 TRUE     26 FALSE    17 TRUE         85 FALSE  
19 D     D        22 TRUE     29 TRUE     26 FALSE    16 TRUE         93 FALSE  
20 D     Total   101 FALSE    78 FALSE   115 FALSE    91 FALSE       385 FALSE  
21 Total A        95 FALSE    90 FALSE   111 FALSE    94 FALSE       390 FALSE  
22 Total B        31 FALSE   135 FALSE   114 FALSE    81 FALSE       361 FALSE  
23 Total C       145 FALSE    80 FALSE    87 FALSE    83 FALSE       395 FALSE  
24 Total D        88 FALSE    80 FALSE    88 FALSE    87 FALSE       343 FALSE  
25 Total Total   359 FALSE   385 FALSE   400 FALSE   345 FALSE      1489 FALSE  

sum(final_aggregate$highs_1) 
# 34 suppressions in table

# Run alternate solver using GLPK 
## Requires many more iterations and takes 36mins! Bad!
tic()
final_aggregate <- final_aggregate %>%
  dplyr::mutate(glpk = determine_cell_suppression(n, map$nullspace, solver = "glpk", max_iter = 300))
toc()

convert_tabular(final_aggregate, A2 + B2 ~ C2 + .Measure, measures = c(No. = "n", s = "glpk")) %>%
  print(n = 25)

# However, it appears to have solved the solution optimally
   A2    B2    A_No. A_s   B_No. B_s   C_No. C_s   D_No. D_s   Total_No. Total_s
   <chr> <chr> <int> <lgl> <int> <lgl> <int> <lgl> <int> <lgl>     <int> <lgl>  
 1 A     A         1 TRUE     35 TRUE     37 FALSE    31 FALSE       104 FALSE  
 2 A     B         2 TRUE     36 FALSE    27 FALSE     9 TRUE         74 FALSE  
 3 A     C        38 FALSE    37 FALSE    24 FALSE    29 FALSE       128 FALSE  
 4 A     D        22 FALSE    25 TRUE     35 FALSE     2 TRUE         84 FALSE  
 5 A     Total    63 FALSE   133 FALSE   123 FALSE    71 FALSE       390 FALSE  
 6 B     A        40 TRUE     30 TRUE     31 FALSE     6 FALSE       107 FALSE  
 7 B     B         3 TRUE     39 FALSE    28 FALSE    24 TRUE         94 FALSE  
 8 B     C        32 FALSE    15 TRUE     18 FALSE     2 TRUE         67 FALSE  
 9 B     D        35 FALSE    10 FALSE    10 FALSE    32 FALSE        87 FALSE  
10 B     Total   110 FALSE    94 FALSE    87 FALSE    64 FALSE       355 FALSE  
11 C     A        21 FALSE     6 FALSE     9 FALSE    24 FALSE        60 FALSE  
12 C     B        16 FALSE    36 FALSE    30 FALSE    23 FALSE       105 FALSE  
13 C     C        39 FALSE    22 FALSE    19 FALSE    35 FALSE       115 FALSE  
14 C     D         9 FALSE    16 FALSE    17 FALSE    37 FALSE        79 FALSE  
15 C     Total    85 FALSE    80 FALSE    75 FALSE   119 FALSE       359 FALSE  
16 D     A        33 FALSE    19 FALSE    34 FALSE    33 FALSE       119 FALSE  
17 D     B        10 FALSE    24 FALSE    29 FALSE    25 FALSE        88 FALSE  
18 D     C        36 FALSE     6 TRUE     26 FALSE    17 TRUE         85 FALSE  
19 D     D        22 FALSE    29 TRUE     26 FALSE    16 TRUE         93 FALSE  
20 D     Total   101 FALSE    78 FALSE   115 FALSE    91 FALSE       385 FALSE  
21 Total A        95 FALSE    90 FALSE   111 FALSE    94 FALSE       390 FALSE  
22 Total B        31 FALSE   135 FALSE   114 FALSE    81 FALSE       361 FALSE  
23 Total C       145 FALSE    80 FALSE    87 FALSE    83 FALSE       395 FALSE  
24 Total D        88 FALSE    80 FALSE    88 FALSE    87 FALSE       343 FALSE  
25 Total Total   359 FALSE   385 FALSE   400 FALSE   345 FALSE      1489 FALSE 

sum(final_aggregate$glpk)
# Only 16 suppressions

# Here's the weird part! If I run the default solver again after GLPK, HiGHS gets the same solution
final_aggregate <- final_aggregate %>%
  dplyr::mutate(highs_2 = determine_cell_suppression(n, map$nullspace))

identical(final_aggregate$glpk, final_aggregate$highs_2)
# TRUE

sum(final_aggregate$highs_2)
# 16
```

What could cause HiGHS/ROI to act in this manner? Could ROI or some downstream C++ code have a shared cache causing issues??? Running the HiGHS solver in a fresh session many times produces the same result every time. As soon as I run GLPK and it solves correctly, if I run HiGHS again, it suddenly can find the optimal solution. While we can leave GLPK as the default solver, 36mins to solve a 3 dimensional table with four elements each is a bit excessive when the HiGHS solver is done in 3.5s (albeit in rare cases like in the above is suboptimal/incorrect). I don't love leaving a solver as default that ends on a (very) suboptimal solution, especially not knowing the incidence of these errors. I'm sure there must be something wrong in the code somewhere - HiGHS is a highly regarded solver and it seems odd that it is finding such a suboptimal 'solution' that still has issues (34 vs 16 suppressions with suppressed weight of 723 vs 238), and there is such a vast difference in the number of `max_iter`/constraints added.

I have reviewed the nullspace and it appears correct (although it is large so I didn't go line by line) - besides why would GLPK solve correctly if it was wrong? I have also built the mapping manually rather than the code I did above in case I did something weird in there but no dice. The bug is replicable on another machine although interestingly while GLPK had a more optimised solution, it also had a back-calculation issue similar to the HiGHS solution above where one of the secondary suppressions can be recalculated).

It is doing my head in! 🤯

**Comment by PeterM74** [comment]
Ok, I was going to take a break from this but I thought I would do one last attempt at a solution. I figured (part) of it out and now the HiGHS solver is giving the same (likely) optimal solution as GLPK! The issue is here:

```as.logical(master_soln$solution)```

At a certain depth of table, it stops returning 0/1 integers and starts returning floating point numbers that are 1 or near 0 (e.g. 2.1E-15). The problem? `as.logical` returns any non-zero number as TRUE:

```
as.logical(c(0, 2e-15, 1e-10, 0.9, 1))
[1] FALSE  TRUE  TRUE  TRUE  TRUE
```

When I printed the output of master_soln, HiGHS was returning several near-zero values which were being evaluated as suppressions. They weren't supposed to be suppressions and the attacker subproblems kept being resolved despite there (appearing) to be unresolved suppressions within some of the dimensions (because cells were being suppressed that were irrelevant to the subproblem). Gah! Please let that be the last problem I find! I am so exhausted 🤣

It still doesn't solve why switching between HiGHS and GLPK resulted in differing solutions though! However maybe this is something we can raise in the vignette as a warning of potential cache issues when switching between solvers.

**Comment by inpowell** [comment]
Excellent detective work, and I’m glad you were able to find a solution. It looks like it’s intended behaviour for HiGHS to sometimes return a not-quite-integer solution for an integer problem ([HiGHS/#1626](https://github.com/ERGO-Code/HiGHS/issues/1626)). So, it would be quite reasonable to round the solution before conversion.

**Comment by PeterM74** [comment]
Thanks Ian, I had thrown in the towel when I asked for you to take a look but thankfully I gave it one final go and thought to review the actual solution object before it is coerced as logical. I have tested across 40 RNG tables and they all look good to me so have submitted PR #18 with these changes. No rush as I know you are busy with lots of life admin/changes at the moment! I appreciate you taking a look and providing the HiGHS integer context.
